### PR TITLE
Fix array_concat with NULL arrays

### LINKED ISSUE #16349: array_concat fails with when passed NULL list literals
### Describe the bug

The array_concat function throws an Arrow compute error `concat requires input of at least one array` when called with NULL list literals (e.g., NULL::integer[]). The function incorrectly attempts to pass empty arrays to Arrow's concat function instead of recognizing that NULL list literals should be handled as null inputs.

> Note: Mixed cases work correctly (e.g., array_concat(NULL::integer[], [1,2,3]) returns [1,2,3]), so the issue is specifically when ALL arguments are NULL arrays.

### To Reproduce

```
-- This query should return NULL but throws an error instead
SELECT array_concat(NULL::integer[]);

-- These variations also fail:
SELECT array_concat(NULL::integer[], NULL::integer[]);
SELECT array_concat(NULL::text[]);
SELECT array_concat(NULL::bigint[]);
```

### Expected behavior

The function should handle NULL list literals gracefully and return NULL, similar to how other array functions handle NULL inputs:
```
-- Should return:
SELECT array_concat(NULL::integer[]);
-- Expected: NULL

SELECT array_concat(NULL::integer[], NULL::integer[]);
-- Expected: NULL
```

### Additional context

DataFusion version: v48.0.0 (current main branch)
Rust version: 1.87.0

### PR DESCRIPTION
**Pull request by alexanderbianchi**
Fixes issue where 'select array_concat(NULL::integer[])' would throw 'Arrow error: Compute error: concat requires input of at least one array'

## Which issue does this PR close?

- Closes #16349 

## What changes are included in this PR?

- Fix array_concat_inner to properly handle NULL list arrays by checking null_count
- Add logic to create properly typed empty arrays for all-null inputs
- Support NULL handling for List, LargeList, and FixedSizeList types
- Add comprehensive test coverage for NULL array scenarios

### CODE REVIEW

File: datafusion/functions-nested/src/concat.rs [hunk 1]
```diff
@@ -361,15 +362,44 @@ pub(crate) fn array_concat_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+            }
+            DataType::List(_) | DataType::FixedSizeList(..) => {
+                // Check if this list array has any non-null rows
+                if arg.null_count() < arg.len() {
+                    all_null = false;
+                }
+            }
+            _ => {
+                all_null = false;
+            }
```
Comment (gabotechs) [inline datafusion/functions-nested/src/concat.rs]: Nice, do you think the null count check could be performed always regarding of the type? something like:
```rust
        match arg.data_type() {
            DataType::Null => continue,
            DataType::LargeList(_) => large_list = true,
            _ => (),
        }
+       if arg.null_count() < arg.len() {
+            all_null = false;
+       }
```
?

File: datafusion/functions-nested/src/concat.rs [hunk 2]
```diff
@@ -361,15 +362,44 @@ pub(crate) fn array_concat_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+                GenericListArray::<i32>::new_null(field.clone(), args[0].len()),
+            )),
+            Some(DataType::LargeList(field)) => Ok(Arc::new(
+                GenericListArray::<i64>::new_null(field.clone(), args[0].len()),
+            )),
+            Some(DataType::FixedSizeList(field, size)) => Ok(Arc::new(
+                FixedSizeListArray::new_null(field.clone(), *size, args[0].len()),
+            )),
+            _ => Ok(Arc::new(NullArray::new(args[0].len()))),
+        }
```
Comment (gabotechs) [inline datafusion/functions-nested/src/concat.rs]: There's this `ArrayData::new_null()` method that has already a match statement similar to this, maybe it can save some lines, something like:
```rust
    if all_null {
+        Ok(arrow::array::make_array(ArrayData::new_null(
+            args[0].data_type(),
+            args[0].len(),
+        )))
    } else if large_list {
```

File: datafusion/sqllogictest/test_files/array.slt [hunk 3]
```diff
@@ -3070,6 +3070,30 @@ select array_concat([]);
 ----
 []
 
+# test with NULL array
+query ?
+select array_concat(NULL::integer[]);
+----
+NULL
+
```
Comment (gabotechs) [inline datafusion/sqllogictest/test_files/array.slt:3078]: How about adding a couple of cases that test the other different list types?

```sql
# test with NULL LargeList
query ?
select array_concat(arrow_cast(NULL::string[], 'LargeList(Utf8)'));
----
NULL

# test with NULL FixedSizeList
query ?
select array_concat(arrow_cast(NULL::string[], 'FixedSizeList(2, Utf8)'));
----
NULL

```

**Review by gabotechs — APPROVED** [review]
💯 awesome, cc @alamb this is good for a review
